python/ccf/cose.py

```diff
@@ -124,20 +127,78 @@ def create_cose_sign1(
     return msg.encode()
 
 
-DESCRIPTION = """Create and sign a COSE Sign1 message for CCF governance
+def create_cose_sign1_prepare(
+    payload: bytes,
+    cert_pem: Pem,
+    additional_headers: Optional[dict] = None,
+) -> dict:
+    cert = load_pem_x509_certificate(cert_pem.encode("ascii"), default_backend())
+    alg = default_algorithm_for_key(cert.public_key())
+    kid = cert_fingerprint(cert_pem)
+
+    headers = {pycose.headers.Algorithm: alg, pycose.headers.KID: kid}
+    headers.update(additional_headers or {})
+    msg = Sign1Message(phdr=headers, payload=payload)
+    tbs = cbor2.dumps(["Signature1", msg.phdr_encoded, b"", payload])
+
+    assert cert.signature_hash_algorithm
+    digester = hashes.Hash(cert.signature_hash_algorithm)
+    digester.update(tbs)
+    digest = digester.finalize()
+    return {"alg": alg, "value": base64.b64encode(digest).decode()}
+
+
+def create_cose_sign1_finish(
+    payload: bytes,
+    cert_pem: Pem,
+    signature: str,
+    additional_headers: Optional[dict] = None,
+) -> bytes:
+    cert = load_pem_x509_certificate(cert_pem.encode("ascii"), default_backend())
+    alg = default_algorithm_for_key(cert.public_key())
+    kid = cert_fingerprint(cert_pem)
+
+    headers = {pycose.headers.Algorithm: alg, pycose.headers.KID: kid}
```

Comment:
Re-encoding the protected header works here because it's using the exact same dict with the same ordering of fields between `prepare` and `finish` but at a minimum there should be a comment on that. It may be better to create a COSE message without signature, and then just append the signature.

Reply:
I also considered that, but the UX benefit was unclear: in one case that's an extra file to save and re-provide, in the other case it's "call with the exact same arguments again", but pass the signature this time. Ultimately the latter was closer to the existing procedure, so I stuck with that.

Reply:
After discussion with @eddyashton I will add that as a feature in a later PR, but for now we will go with this.